Workflow guidelines version 1

SimonHeybrock opened the pull request:
Add initial version of workflow guidelines.

jl-wynen commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+
+**Note**
+Most of this is avoid by following S.2.
+A bad exampleAvoid would be to write to total raw counts to the output metadata, as this would require keeping the large data alive in memory, unless it is ensured that the task runs early.
```
```suggestion
A bad example would be writing the total raw counts to the output metadata, as this would require keeping the large data alive in memory, unless it is ensured that the task runs early.
```

jl-wynen commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+
+**Note**
+Most of this is avoid by following S.2.
+A bad exampleAvoid would be to write to total raw counts to the output metadata, as this would require keeping the large data alive in memory, unless it is ensured that the task runs early.
```
But how do we do this instead? This note just forbids writing certain things as metadata. But that cannot be avoided, e.g., in reflectometry as we need to write wavelength and theta ranges.

SimonHeybrock replied:
There will be exceptions, not everything can be "avoided", this is why I chose this formulation instead of making it stricter.

jl-wynen replied:
Then I'd like clarification about how strict these guidelines are. As it is written, it sounds like a command. So if this one may be violated, can all rules be broken if we can argue for it?

SimonHeybrock replied:
Time will tell? I think we should be agile and avoid over-analyzing for now, see what works in practice and what does not, then iterate and improve.

SimonHeybrock replied:
Also, in the case you mentioned, shouldn't the wavelength and theta ranges report the *used* ranges, instead of those in the raw data? I presume the workflow may limit this, and in any case it can be computed from derived data instead of raw data, limiting the problem.

jl-wynen replied:
> Also, in the case you mentioned, shouldn't the wavelength and theta ranges report the _used_ ranges, instead of those in the raw data? I presume the workflow may limit this, and in any case it can be computed from derived data instead of raw data, limiting the problem.

Possibly. The current workflow does not allow limiting theta. Also, what if the bounds are much larger than the data? I'd want to report the limits of the data that is actually used instead.

jl-wynen commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+- Required for subsequent filtering, unless part of the workflow.
+
+**Note**
+There should be a workflow parameters (flag) to select whether to return event data or not.
```
```suggestion
There should be a workflow parameter (flag) to select whether to return event data or not.
```

jl-wynen commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+
+**Example**
+- If time-of-flight is single-precision, wavelength and momentum transfer should be single-precision.
+- In counts are single-precision, reduced intensity should be single-precision.
```
```suggestion
- If counts are single-precision, reduced intensity should be single-precision.
```

jl-wynen commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+Most of this is avoid by following S.2.
+A bad exampleAvoid would be to write to total raw counts to the output metadata, as this would require keeping the large data alive in memory, unless it is ensured that the task runs early.
+
+### S.4: Handle single- and double-precision transparently
```
I don't think 'transparent' is the right word here. How about 'Maintain floating point precision'?

jl-wynen commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+### S.6: Propagation of uncertainties in broadcast operations should support "drop" and "upper-bound" strategies, "upper-bound" shall be the default
+
+**Reason**
+Unless explicitly computed, the exact propagation of uncertainties in broadcast operations is not well-defined.
```
It is well defined but computationally intractable.

jl-wynen commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+**Reason**
+Unless explicitly computed, the exact propagation of uncertainties in broadcast operations is not well-defined.
+Dropping uncertainties is not desireable in general, as it may lead to underestimation of the uncertainties, but we realize that the upper-bound approach may not be suitable in all cases.
+We should therefore support two strategies, "drop" and "upper-bound", and "upper-bound" should be the default.
```
Is this really something the user should toggle? How can the upper-bound strategy be valid for some data but not other for the same workflow?

SimonHeybrock replied:
I am pretty sure that at least the IDS or IS will want to toggle this.

jl-wynen commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+Providers should be side-effect free, and should not write files or make write requests to services.
+
+**Note**
+Workflow may run many times, or in parallel, or task may be retried after failure, and we want to avoid side-effects in these cases.
```
```suggestion
Workflows may run many times, or in parallel, or tasks may be retried after failure, and we want to avoid side-effects in these cases.
```

jl-wynen commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+See [Systematic underestimation of uncertainties by widespread neutron-scattering data-reduction software](http://dx.doi.org/10.3233/JNR-220049) for a discussion of the topic.
+TODO Add reference to upper-bound approach.
+
+### S.7: Do not write files or make write requests to services such as SciCat in providers
```
Not sure we should keep SciCat requests out of the pipeline. This adds additional code around the pipeline and removed the relevant parameters from the purview of the pipeline (it can only track local filenames which are not unique instead of unique SciCat IDs)

SimonHeybrock replied:
I was specifically referring to write-requests, would you want to much such requests?

jl-wynen replied:
Somehow I did not register the 'write'. Never mind then

jl-wynen commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+Referencing the guideline ID in the test name makes it easier to find the relevant guideline (or vice-versa), or remove the test if the guideline is removed.
+
+**Note**
+Note all guidelines are testable.
```
```suggestion
Not all guidelines are testable.
```

jl-wynen commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+
+**Reason** Unit tests for providers are easier to write and maintain than for entire workflows.
+
+**Note** This does not mean that we should not write tests for entire workflows, but that we should also write tests for providers.
```
How about a guideline for structuring the test directory? And for how to select whether to run integration tests or not. If they take a long time to run, it would be useful to be able to easily deselect them.

SimonHeybrock replied:
I think splitting unit tests from integration tests (or something like that) is a great suggestion! As we have not done that before, maybe we can add it in a future version, once we have tried it out in a project?

nvaytet commented on docs/user-guide/reduction-workflow-guidelines.md:
```diff
@@ -0,0 +1,139 @@
+# Reduction Workflow Guidelines
```
Nothing to add from my side :+1:

SimonHeybrock commented:
@jl-wynen @nvaytet I think I addressed all comments. I have cleaned this up and suggest releasing a first agreed-upon version such that we can begin implementation. I included a listing of the remaining items which we can add subsequently.
